crates/electrum/src/bdk_electrum_client.rs

```diff
@@ -414,55 +444,166 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+        for height in heights {
+            headers.insert(height, self.fetch_header(height)?);
+        }
+
+        // Validate proofs.
+        for (txid, merkle_res) in proofs {
+            let height = merkle_res.block_height as u32;
+            let header = headers.get(&height).unwrap();
+
             let mut is_confirmed_tx = electrum_client::utils::validate_merkle_proof(
                 &txid,
                 &header.merkle_root,
                 &merkle_res,
             );
 
-            // Merkle validation will fail if the header in `block_header_cache` is outdated, so we
-            // want to check if there is a new header and validate against the new one.
+            // Retry with updated header if validation fails.
             if !is_confirmed_tx {
-                header = self.update_header(merkle_res.block_height as u32)?;
+                let updated_header = self.update_header(height)?;
+                headers.insert(height, updated_header);
                 is_confirmed_tx = electrum_client::utils::validate_merkle_proof(
                     &txid,
-                    &header.merkle_root,
+                    &updated_header.merkle_root,
                     &merkle_res,
                 );
             }
 
             if is_confirmed_tx {
+                let header = headers.get(&height).unwrap();
                 tx_update.anchors.insert((
                     ConfirmationBlockTime {
                         confirmation_time: header.time as u64,
                         block_id: BlockId {
-                            height: merkle_res.block_height as u32,
+                            height,
                             hash: header.block_hash(),
                         },
                     },
                     txid,
                 ));
             }
         }
+
         Ok(())
     }
 
+    /// Remove any proofs for blocks that may have been re-orged out.
+    ///
+    /// Checks if the latest cached block hash matches the current chain tip. If not, evicts proofs
+    /// for blocks that were re-orged out, stopping at the fork point.
+    fn clear_stale_proofs(&self) -> Result<(), Error> {
+        let mut cache = self.merkle_cache.lock().unwrap();
+
+        // Collect one (height, old_hash) pair per proof.
+        let mut entries: Vec<(u32, BlockHash)> = cache
+            .iter()
+            .map(|((_, old_hash), res)| (res.block_height as u32, *old_hash))
+            .collect();
+
+        // Sort descending and dedup so we only check each height once.
+        entries.sort_unstable_by(|a, b| b.0.cmp(&a.0));
+        entries.dedup();
+
+        // Evict any stale proofs until fork point is found.
+        for (height, old_hash) in entries {
+            let current_hash = self.fetch_header(height)?.block_hash();
+            if current_hash == old_hash {
+                break;
+            }
+            cache.retain(|&(_txid, bh), _| bh != old_hash);
+        }
+        Ok(())
+    }
```

Comment:
Reorgs don't happen that often so we won't have much "extra data". This method looks like it's `O(n^2)`. Let's remove it.